### PR TITLE
Improving the ROCm build process.

### PR DESCRIPTION
**Pull request by Jorghi12**
**### Fix 1)**
The NVCC compiler automatically pre includes CUDA-specific header files when handling code in cu/cuh files (e.g. --preinclude "cuda_runtime.h"). The HCC compiler didn't do this automatically for us, so we need to add this logic into the transpilation script.

Source: http://www.cs.cmu.edu/afs/cs/academic/class/15668-s11/www/cuda-doc/nvcc.pdf

**### Fix 2)**
Also, I amended the transpiler with a fix for dynamic shared memory allocation when kernels are launched from the host side. These warnings will no longer show up on the console anymore. 
https://github.com/ROCm-Developer-Tools/HIP/blob/master/docs/markdown/hip_kernel_language.md#__shared__

**### Fix 3)**
We need to include the SHARED flag when calling hip_add_library. There is logic in FindHIP.CMake such that if this flag is missing, then the -fPIC flag will not be passed into HCC while linking.  
https://github.com/ROCm-Developer-Tools/HIP/blob/master/cmake/FindHIP.cmake#L398

**### Fix 4)**
Compile Definitions that were set in the CMakeLists directory were being dropped when building with the FindHIP.CMake toolchain. I've added the workaround logic in order to handle this issue.

### CODE REVIEW

File: tools/amd_build/pyHIPIFY/hipify-python.py [hunk 1]
```diff
@@ -450,8 +450,9 @@ def disable_asserts(input_string):
         output_string = output_string.replace(input_string[start:p_end + 1], "")
     return output_string
 
+
```
Comment (iotamudelta) [inline tools/amd_build/pyHIPIFY/hipify-python.py]: Added line should be removed, no?

File: tools/amd_build/pyHIPIFY/hipify-python.py [hunk 2]
```diff
@@ -471,6 +473,53 @@ def replace_math_functions(input_string):
     output_string = re.sub("std::pow\(", "::pow(", output_string)
     return output_string
 
+
+def replace_extern_shared(input_string):
```
Comment (iotamudelta) [inline tools/amd_build/pyHIPIFY/hipify-python.py]: Good catch! :-)

File: tools/amd_build/pyHIPIFY/hipify-python.py [hunk 3]
```diff
@@ -471,6 +473,53 @@ def replace_math_functions(input_string):
     output_string = re.sub("std::pow\(", "::pow(", output_string)
     return output_string
 
+
```
Comment (iotamudelta) [inline tools/amd_build/pyHIPIFY/hipify-python.py]: remove empty line?

File: tools/amd_build/pyHIPIFY/hipify-python.py [hunk 4]
```diff
@@ -471,6 +473,53 @@ def replace_math_functions(input_string):
...
+    output_string = input_string
+    output_string = re.sub(
+        r"extern\s+([\w\(\)]+)?\s*__shared__\s+([\w:<>\s]+)\s+(\w+)\s*\[\s*\]\s*;",
+        lambda inp: "HIP_DYNAMIC_SHARED({0} {1}, {2})".format(
+            inp.group(1) or "", inp.group(2), inp.group(3)), output_string)
+
+    return output_string
+
+
+def hip_header_magic(input_string):
```
Comment (iotamudelta) [inline tools/amd_build/pyHIPIFY/hipify-python.py]: Looks good - but I haven't seen this issue pop up yet anywhere. Could you comment where you've seen this?

File: tools/amd_build/pyHIPIFY/hipify-python.py (hunk 3 above)
Comment (Jorghi12) [reply tools/amd_build/pyHIPIFY/hipify-python.py]: My IDE's linter loved to scream at me haha.

Apparently it's good style to have two lines between function definitions in PyThon. https://www.python.org/dev/peps/pep-0008/?#blank-lines

File: tools/amd_build/pyHIPIFY/hipify-python.py (hunk 1 above)
Comment (Jorghi12) [reply tools/amd_build/pyHIPIFY/hipify-python.py]: #PEP8

**Comment by iotamudelta** [comment]
As expected the recent update to hipify causes a conflict. Can you resolve? Thanks!

**Pull request closed without merging**